clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchCollector.java

```diff
@@ -327,21 +327,27 @@ private void handleInitializeErrors(final CompletedFetch completedFetch, final E
         final long fetchOffset = completedFetch.nextFetchOffset();
 
         if (error == Errors.NOT_LEADER_OR_FOLLOWER ||
-                error == Errors.REPLICA_NOT_AVAILABLE ||
+                error == Errors.FENCED_LEADER_EPOCH) {
+            log.debug("Error in fetch for partition {}: {}", tp, error.exceptionName());
+            requestMetadataUpdate(metadata, subscriptions, tp);
+        } else if (error == Errors.REPLICA_NOT_AVAILABLE ||
                 error == Errors.KAFKA_STORAGE_ERROR ||
-                error == Errors.FENCED_LEADER_EPOCH ||
                 error == Errors.OFFSET_NOT_AVAILABLE) {
             log.debug("Error in fetch for partition {}: {}", tp, error.exceptionName());
             requestMetadataUpdate(metadata, subscriptions, tp);
+            subscriptions.awaitUpdate(tp);
```

Comment:
With this change, if the replica is not available, we will flag the partition as awaiting a metadata update. Is this a key part of this change? Why don't we want the first `if` block to also await an update?

Reply:
I missed that. For this case, there is no need to await a metadata update.
I think simply initializing the PreferredReadReplica should be enough.
Since `FetchUtils.requestMetadataUpdate()` is already being called, it should also be initializing.